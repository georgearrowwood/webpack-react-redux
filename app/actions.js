import productsApi from './api/products'

export const addProduct = (title) => (dispatch, getState) => {
  return {
    type: 'ADD_PRODUCT',
    title
  }
}

export const requestProducts = () => ({
  type: 'REQUEST_PRODUCTS'
})

export const receiveProducts = (data) => ({
  type: 'RECEIVE_PRODUCTS',
  products: data
})

export const fetchProducts = () => (dispatch, getState) => {
  const state = getState()
  if (state.isFetching) return false
  dispatch(requestProducts())
  productsApi.getList()
    .then((data) => {
      dispatch(receiveProducts(data.data.products))
    })
}
