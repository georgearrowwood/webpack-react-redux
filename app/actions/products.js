import dispatcher from '../utils/dispatcher';
import productConstants from '../constants/products';
import productsApi from '../api/products';

export default {
  /**
   * @param  {string} text
   */
  create: function(product) {
    if (product && product.title) {
      productsApi.create(product)
        .then(result => {
          if (result.data.success) {
            this.fetchList();
          }
        })
    }
  },

  fetchList: function() {
    productsApi.getList()
      .then(result => {
        dispatcher.dispatch({
          actionType: productConstants.PRODUCTS_LIST_DATA_RECEIVED,
          products: result.data.products
        });
      })
  },

  /**
   * @param  {integer} id
   */
  delete: function(id) {
    console.log('d1', id);
    if (Number.isInteger(id)) {
      productsApi.delete(id)
        .then(result => {
          if (result.data.success) {
            this.fetchList();
          }
        })
    }
  }

};
