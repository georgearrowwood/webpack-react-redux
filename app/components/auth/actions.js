import request from '../../modules/request';

export const loginAction = ({ login, password }, history) => (dispatch) => {
    console.log('6', login);
      
    request.post('/api/login', { login, password })
      .then(res => {   
        console.log('res log', res);
        
        dispatch({ type: 'AUTHENTICATED' });
        /* eslint-env browser */
        localStorage.setItem('user-token', res.data.token);
        // history.push('/');
        history.goBack();
      })  
      .catch ((error) => {
        console.log('err log', error);
        dispatch({
          type: 'AUTHENTICATION_ERROR',
          payload: 'Invalid email or password',
        });
      });
};

export const as = 1;

