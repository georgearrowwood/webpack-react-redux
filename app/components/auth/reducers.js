export default (state = {}, action) => {
  console.log('au red', state);
  switch (action.type) {
    case 'AUTHENTICATED':
      return { ...state, authenticated: true };
    case 'UNAUTHENTICATED':
      return { ...state, authenticated: false };
    case 'AUTHENTICATION_ERROR':
      return { ...state, error: action.payload };
    default:
      return state;
  }
};
