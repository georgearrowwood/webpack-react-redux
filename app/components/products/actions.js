import services from './services';

export const requestProducts = () => ({
  type: 'REQUEST_PRODUCTS',
});

export const receiveProducts = data => ({
  type: 'RECEIVE_PRODUCTS',
  products: data,
});

export const fetchProducts = () => (dispatch, getState) => {
  const state = getState();
  if (state.isFetching) return false;

  dispatch(requestProducts());
  return services().getList()
    .then(data => dispatch(receiveProducts(data.data.products)));
};

export const addProduct = data => (dispatch) => {
  services().addOne(data.title)
    .then(() => {
      dispatch(fetchProducts());
    });
};

export const deleteProduct = id => dispatch =>
  services().deleteOne(id)
    .then(() => {
      dispatch(fetchProducts());
    });
