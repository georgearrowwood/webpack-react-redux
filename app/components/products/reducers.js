function products(state = {}, action) {
  switch (action.type) {
    case 'REQUEST_PRODUCTS':
      return {
        ...state,
        isFetching: true,
      };
    case 'RECEIVE_PRODUCTS':
      return {
        ...state,
        isFetching: false,
        items: action.products,
      };
    default:
      return state;
  }
}

export default products;
