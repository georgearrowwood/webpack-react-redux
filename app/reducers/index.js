import { combineReducers } from 'redux'
import { routerReducer } from 'react-router-redux'
import products from './products'


const reducers = combineReducers({
  products,
  router: routerReducer
})

export default reducers
