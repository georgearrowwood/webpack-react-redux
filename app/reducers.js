import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';
import { reducer as formReducer } from 'redux-form';

import auth from './components/auth/reducers';
import products from './components/products/reducers';

const reducers = combineReducers({
  auth,
  products,
  router: routerReducer,
  form: formReducer,
});

export default reducers;
