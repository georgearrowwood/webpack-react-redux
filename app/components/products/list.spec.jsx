import React from 'react';
import { mount } from 'enzyme';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

// import productsStore from './store';
import * as productsActions from './actions';
import ProductsList from './list';
import reducers from './reducers';

const mockProductsList = [
  { id: 1, title: 'test product 1' },
  { id: 2, title: 'test product 2' },
];

describe('ProductsList Container', () => {
  it('Renders component and triggers action fetchList on did mount', () => {
    const store = createStore(combineReducers({ products: reducers }), applyMiddleware(thunk));
    jest.spyOn(productsActions, 'fetchProducts').mockImplementation(() => jest.fn());
    const wrapper = mount(<Provider store={store}><ProductsList /></Provider>);
    expect(wrapper).toMatchSnapshot();
    const component = wrapper.find('ProductsList');
    expect(Array.isArray(component.prop('products'))).toEqual(true);
    expect(component.prop('products').length).toEqual(0);
    expect(productsActions.fetchProducts.mock.calls.length).toEqual(1);
    productsActions.fetchProducts.mockRestore();
  });

  it('set init state with products data', () => {
    const store = createStore(
      combineReducers({ products: reducers }),
      { products: { items: mockProductsList } },
      applyMiddleware(thunk),
    );
    const wrapper = mount(<Provider store={store}><ProductsList /></Provider>);
    expect(wrapper).toMatchSnapshot();
    const component = wrapper.find('ProductsList');
    expect(Array.isArray(component.prop('products'))).toEqual(true);
    expect(component.prop('products').length).toEqual(2);
  });
});
