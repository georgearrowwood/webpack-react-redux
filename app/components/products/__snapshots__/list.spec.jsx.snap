// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList Container Renders component and triggers action fetchList on did mount 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(ProductsList)>
    <ProductsList
      auth={undefined}
      dispatch={[Function]}
      isFetching={undefined}
      products={Array []}
    >
      <ProductsListView
        deleteHandler={[Function]}
        products={Array []}
      >
        <ul />
      </ProductsListView>
    </ProductsList>
  </Connect(ProductsList)>
</Provider>
`;

exports[`ProductsList Container set init state with products data 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(ProductsList)>
    <ProductsList
      auth={undefined}
      dispatch={[Function]}
      isFetching={true}
      products={
        Array [
          Object {
            "id": 1,
            "title": "test product 1",
          },
          Object {
            "id": 2,
            "title": "test product 2",
          },
        ]
      }
    >
      <ProductsListView
        deleteHandler={[Function]}
        products={
          Array [
            Object {
              "id": 1,
              "title": "test product 1",
            },
            Object {
              "id": 2,
              "title": "test product 2",
            },
          ]
        }
      >
        <ul>
          <li
            key="1"
          >
            <span>
              1
               - 
              test product 1
               :
            </span>
            <a
              href="delete/{product.id}"
              onClick={[Function]}
            >
              X
            </a>
          </li>
          <li
            key="2"
          >
            <span>
              2
               - 
              test product 2
               :
            </span>
            <a
              href="delete/{product.id}"
              onClick={[Function]}
            >
              X
            </a>
          </li>
        </ul>
      </ProductsListView>
    </ProductsList>
  </Connect(ProductsList)>
</Provider>
`;
